docs/ADR-43-signed-fetch.md

```diff
@@ -0,0 +1,131 @@
+# Signed Fetch
+
+## Context and Problem Statement
+
+We want to integrate our dApps with the explorer in order to provide a better experience for our users. To do so we need to send request authenticated as a wallet.
+
+## Considered options
+
+### Alt 1
+
+Ask the user to sign each request
+
+#### Pros
+
+- Users will known each time an scene will interact in their name
+- No intermediate code is required
+
+#### Cons
+
+- For privates APIs it will require to many interactions for the users, potentially degrading the experience within the explorer
+
+### Alt 2
+
+Use [`decentraland-crypto`](https://github.com/decentraland/decentraland-crypto) as intermediate to sign each request
+
+#### Pros
+
+- It only requires one sign from the user
+- It is already integrated in the explorer
+  - [`signedFetch`](https://docs.decentraland.org/development-guide/network-connections/#signed-requests)
+- Libraries to interact with these firms are already published
+  - [`decentraland-crypto`](https://github.com/decentraland/decentraland-crypto)
+  - [`decentraland-crypto-middleware`](https://github.com/decentraland/decentraland-crypto-middleware)
+  - [`decentraland-crypto-fetch`](https://github.com/decentraland/decentraland-crypto-fetch)
+
+#### Cons
+
+- It requires an extensive documentation
+- It requires extra code
+
+## Decision
+
+The option to use [`decentraland-crypto`](https://github.com/decentraland/decentraland-crypto) to interact with our services was chosen in order to provide a better experience to our users
+
+## How to sign a request
+
+### Generate a payload
+
+In order to minimize the reutilization of request each payload must include data about the request and the moment it is made
+
+```typescript
+
+/**
+ * Request method
+ */
+const METHOD: "GET" | "POST" | "PUT" | "PATCH" | "DELETE" = `POST`
+
+/**
+ * request path without domain, query string or hash
+ * @example
+ * - `/ping`
+ * - `new URL('https://decentraland.org/ping').pathname`
+ */
+const PATH: string = `/ping`
+
+/**
+ * request timestamp
+ * each service decide how long will consider valid a request using this timestamp,
+ * if timestamp is greater than now the request should fail
+ */
+const TIMESTAMP: number = Date.now()
+
+/**
+ * request metadata
+ * can include extra data about the service that is making the request
+ * shouldn't contains data required to make the request, if it is empty
+ * and empty object should be use
+ */
+const METADATA: string = JSON.stringify({ /* extra data */  })
+
+/**
+ * Payload
+ */
+const payload = (
+  [
+    METHOD,
+    PATH,
+    TIMESTAMP,
+    METADATA
```

Comment:
Can we take the chance, and use this ADR to review the sign algorithm?

It seems like the current approach is vulnerable to at least four attacks:
- Reuse signature with different `body` content
- Reuse signature with different `queryString`
- Reuse signature with different `headers`
- Reuse signature with different `origin` or `host`

I'd like to propose the following changes to the signed payload:

1. Prepend the version of the signer algorithm, i.e. `sig1`
2. Append the `queryString` to the signature list (payload in this algorithm)
3. Append the sha256 of the body payload if present, otherwise append empty string
4. Append a list of signed headers and their value
5. Replace the `.join(':')` by `.join('\n')` to account for headers and JSON encoded new lines `\\n`
6. Remove the `.toLowerCase()` of the result since it lowers the security of the algorithm

All of the suggestions take into account what the AwsSignatureAlgorithmV4 does https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html

I think we could gather in a call to evaluate the best way to do this to preserve backwards compatibility and make all the changes only once.

Reply:
I'd also suggest to build on giant'shoulders and use the same encoding as they do, since it is battle tested. We can leverage some libraries to do so like https://github.com/RBrNx/agnostic-aws-signature/blob/master/src/index.js or any of the following https://www.npmjs.com/search?q=signature%20v4

Reply:
Somebody took the initiative seriously https://httpwg.org/http-extensions/draft-ietf-httpbis-message-signatures.html